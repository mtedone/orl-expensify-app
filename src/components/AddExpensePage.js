import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

// Refactors dispatch(addExpense) here so the component is more testable
const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)),
});

// First argument is mapStateToProps, the second is mapDispatchToProps
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
