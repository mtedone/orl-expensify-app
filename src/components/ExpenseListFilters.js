import React from 'react';
import { v4 as uuid4 } from 'uuid';
import { connect } from 'react-redux';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({
      calendarFocused,
    }));
  };

  onTextChange = (e) => {
    // dispatch is available from components
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    const val = e.target.value;

    if (val === 'date') {
      this.props.sortByDate();
    } else if (val === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId={'startDateId99098098098098'}
          endDate={this.props.filters.endDate}
          endDateId={'endDateId90df8accj08rewff'}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          displayFormat="DD MMM YYYY"
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
