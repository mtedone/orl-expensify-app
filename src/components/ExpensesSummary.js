import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => (
  <div>
    {props.expensesCount === 1 && (
      <p>
        Viewing 1 expense totalling
        {' ' + numeral(props.expensesTotal / 100).format('$0,0.00')}
      </p>
    )}
    {props.expensesCount > 1 && (
      <p>
        Viewing {props.expensesCount} expenses totalling{' '}
        {' ' + numeral(props.expensesTotal / 100).format('$0,0.00')}
      </p>
    )}
  </div>
);

const mapStateToProps = (state, props) => ({
  expensesCount: state.expenses.length,
  expensesTotal: selectExpensesTotal(state.expenses),
});

export default connect(mapStateToProps)(ExpensesSummary);
