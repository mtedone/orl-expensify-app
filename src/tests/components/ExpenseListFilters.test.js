import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import { ExpenseListFilters } from '../../components/ExpenseListFilters';
import { filters, altFilters } from '../fixtures/filters';
import { setStartDate, setEndDate } from '../../actions/filters';

let setTextFilterSpy,
  sortByDateSpy,
  sortByAmountSpy,
  setStartDateSpy,
  setEndDateSpy,
  wrapper;

beforeEach(() => {
  setTextFilterSpy = jest.fn();
  sortByAmountSpy = jest.fn();
  sortByDateSpy = jest.fn();
  setStartDateSpy = jest.fn();
  setEndDateSpy = jest.fn();
  wrapper = shallow(
    <ExpenseListFilters
      filters={filters}
      setTextFilter={setTextFilterSpy}
      sortByDate={sortByDateSpy}
      sortByAmount={sortByAmountSpy}
      setStartDate={setStartDateSpy}
      setEndDate={setEndDateSpy}
    />
  );
});

test('should render ExpenseListFilters correctly', () => {
  expect(wrapper.debug()).toMatchSnapshot();
});

test('should render ExpenseListFilters with alt data correctly', () => {
  wrapper.setProps({
    filters: altFilters,
  });
  expect(wrapper.debug()).toMatchSnapshot();
});

test('should handle text change', () => {
  wrapper.find('input').simulate('change', {
    target: {
      value: 'bill',
    },
  });
  expect(setTextFilterSpy).toHaveBeenLastCalledWith('bill');
});

test('should sort by date', () => {
  const value = 'date';
  wrapper.setProps({
    filters: altFilters,
  });
  wrapper.find('select').simulate('change', {
    target: { value },
  });
  expect(sortByDateSpy).toHaveBeenCalled();
});

test('should sort by amount', () => {
  const value = 'amount';
  wrapper.find('select').simulate('change', {
    target: { value },
  });
  expect(sortByAmountSpy).toHaveBeenCalled();
});

test('should handle date changes', () => {
  const startDate = moment(0).add(4, 'years');
  const endDate = moment(0).add(8, 'years');
  wrapper.find('withStyles(DateRangePicker)').prop('onDatesChange')({
    startDate,
    endDate,
  });
  expect(setStartDateSpy).toHaveBeenLastCalledWith(startDate);
  expect(setEndDateSpy).toHaveBeenLastCalledWith(endDate);
});

test('should handle date focus changes', () => {
  const calendarFocused = 'endDate';

  wrapper.find('withStyles(DateRangePicker)').prop('onFocusChange')(
    calendarFocused
  );
  expect(wrapper.state('calendarFocused')).toBe(calendarFocused);
});
