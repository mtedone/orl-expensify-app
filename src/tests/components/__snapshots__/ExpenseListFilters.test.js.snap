// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListFilters correctly 1`] = `
"<div>
  <input type=\\"text\\" value=\\"\\" onChange={[Function]} />
  <select value=\\"date\\" onChange={[Function]}>
    <option value=\\"date\\">
      Date
    </option>
    <option value=\\"amount\\">
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker) startDate={{...}} startDateId=\\"startDateId99098098098098\\" endDate={{...}} endDateId=\\"endDateId90df8accj08rewff\\" onDatesChange={[Function]} focusedInput={{...}} onFocusChange={[Function]} displayFormat=\\"DD MMM YYYY\\" numberOfMonths={1} isOutsideRange={[Function: isOutsideRange]} showClearDates={true} startDatePlaceholderText=\\"Start Date\\" endDatePlaceholderText=\\"End Date\\" startDateAriaLabel={[undefined]} endDateAriaLabel={[undefined]} startDateOffset={[undefined]} endDateOffset={[undefined]} disabled={false} required={false} readOnly={false} screenReaderInputMessage=\\"\\" showDefaultInputIcon={false} inputIconPosition=\\"before\\" customInputIcon={{...}} customArrowIcon={{...}} customCloseIcon={{...}} noBorder={false} block={false} small={false} regular={false} keepFocusOnInput={false} renderMonthText={{...}} renderWeekHeaderElement={{...}} orientation=\\"horizontal\\" anchorDirection=\\"left\\" openDirection=\\"down\\" horizontalMargin={0} withPortal={false} withFullScreenPortal={false} appendToBody={false} disableScroll={false} initialVisibleMonth={{...}} keepOpenOnDateSelect={false} reopenPickerOnClearDates={false} renderCalendarInfo={{...}} calendarInfoPosition=\\"bottom\\" hideKeyboardShortcutsPanel={false} daySize={39} isRTL={false} firstDayOfWeek={{...}} verticalHeight={{...}} transitionDuration={[undefined]} verticalSpacing={22} horizontalMonthPadding={[undefined]} dayPickerNavigationInlineStyles={{...}} navPosition=\\"navPositionTop\\" navPrev={{...}} navNext={{...}} renderNavPrevButton={{...}} renderNavNextButton={{...}} onPrevMonthClick={[Function: onPrevMonthClick]} onNextMonthClick={[Function: onNextMonthClick]} onClose={[Function: onClose]} renderCalendarDay={[undefined]} renderDayContents={{...}} renderMonthElement={{...}} minimumNights={1} enableOutsideDays={false} isDayBlocked={[Function: isDayBlocked]} isDayHighlighted={[Function: isDayHighlighted]} minDate={[undefined]} maxDate={[undefined]} monthFormat=\\"MMMM YYYY\\" weekDayFormat=\\"dd\\" phrases={{...}} dayAriaLabelFormat={[undefined]} />
</div>"
`;

exports[`should render ExpenseListFilters with alt data correctly 1`] = `
"<div>
  <input type=\\"text\\" value=\\"bills\\" onChange={[Function]} />
  <select value=\\"amount\\" onChange={[Function]}>
    <option value=\\"date\\">
      Date
    </option>
    <option value=\\"amount\\">
      Amount
    </option>
  </select>
  <withStyles(DateRangePicker) startDate={{...}} startDateId=\\"startDateId99098098098098\\" endDate={{...}} endDateId=\\"endDateId90df8accj08rewff\\" onDatesChange={[Function]} focusedInput={{...}} onFocusChange={[Function]} displayFormat=\\"DD MMM YYYY\\" numberOfMonths={1} isOutsideRange={[Function: isOutsideRange]} showClearDates={true} startDatePlaceholderText=\\"Start Date\\" endDatePlaceholderText=\\"End Date\\" startDateAriaLabel={[undefined]} endDateAriaLabel={[undefined]} startDateOffset={[undefined]} endDateOffset={[undefined]} disabled={false} required={false} readOnly={false} screenReaderInputMessage=\\"\\" showDefaultInputIcon={false} inputIconPosition=\\"before\\" customInputIcon={{...}} customArrowIcon={{...}} customCloseIcon={{...}} noBorder={false} block={false} small={false} regular={false} keepFocusOnInput={false} renderMonthText={{...}} renderWeekHeaderElement={{...}} orientation=\\"horizontal\\" anchorDirection=\\"left\\" openDirection=\\"down\\" horizontalMargin={0} withPortal={false} withFullScreenPortal={false} appendToBody={false} disableScroll={false} initialVisibleMonth={{...}} keepOpenOnDateSelect={false} reopenPickerOnClearDates={false} renderCalendarInfo={{...}} calendarInfoPosition=\\"bottom\\" hideKeyboardShortcutsPanel={false} daySize={39} isRTL={false} firstDayOfWeek={{...}} verticalHeight={{...}} transitionDuration={[undefined]} verticalSpacing={22} horizontalMonthPadding={[undefined]} dayPickerNavigationInlineStyles={{...}} navPosition=\\"navPositionTop\\" navPrev={{...}} navNext={{...}} renderNavPrevButton={{...}} renderNavNextButton={{...}} onPrevMonthClick={[Function: onPrevMonthClick]} onNextMonthClick={[Function: onNextMonthClick]} onClose={[Function: onClose]} renderCalendarDay={[undefined]} renderDayContents={{...}} renderMonthElement={{...}} minimumNights={1} enableOutsideDays={false} isDayBlocked={[Function: isDayBlocked]} isDayHighlighted={[Function: isDayHighlighted]} minDate={[undefined]} maxDate={[undefined]} monthFormat=\\"MMMM YYYY\\" weekDayFormat=\\"dd\\" phrases={{...}} dayAriaLabelFormat={[undefined]} />
</div>"
`;
